compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ReachabilityAnalyzer.java

```diff
@@ -154,393 +140,407 @@ public static ReachabilityAnalyzer getInstance(CompilerContext context) {
+                data.statementReturnsPanicsOrFails = ifStmtReturnsPanicsOrFails && data.statementReturnsPanicsOrFails;
+                data.errorThrown = currentErrorThrown && data.errorThrown;
+                data.breakAsLastStatement = ifStmtBreakAsLastStatement && data.breakAsLastStatement;
+                data.continueAsLastStatement = ifStmtContinueAsLastStatement && data.continueAsLastStatement;
             }
         }
     }
 
     @Override
-    public void visit(BLangDo doNode) {
-        boolean failureHandled = this.failureHandled;
-        checkStatementExecutionValidity(doNode);
-        if (!this.failureHandled) {
-            this.failureHandled = doNode.onFailClause != null;
+    public void visit(BLangDo doNode, AnalyzerData data) {
+        boolean failureHandled = data.failureHandled;
+        checkStatementExecutionValidity(doNode, data);
+        if (!data.failureHandled) {
+            data.failureHandled = doNode.onFailClause != null;
         }
-        analyzeReachability(doNode.body, env);
-        this.failureHandled = failureHandled;
-        analyzeOnFailClause(doNode.onFailClause);
+        analyzeReachability(doNode.body, data);
+        data.failureHandled = failureHandled;
+        analyzeOnFailClause(doNode.onFailClause, data);
     }
 
     @Override
-    public void visit(BLangErrorDestructure errorDestructureStmt) {
-        checkStatementExecutionValidity(errorDestructureStmt);
-        analyzeReachabilityInExpressionIfApplicable(errorDestructureStmt.expr, env);
-        validateAssignmentToNarrowedVariables(getVarRefs(errorDestructureStmt.varRef), errorDestructureStmt.pos, env);
+    public void visit(BLangErrorDestructure errorDestructureStmt, AnalyzerData data) {
+        checkStatementExecutionValidity(errorDestructureStmt, data);
+        analyzeReachabilityInExpressionIfApplicable(errorDestructureStmt.expr, data);
+        validateAssignmentToNarrowedVariables(getVarRefs(errorDestructureStmt.varRef), errorDestructureStmt.pos, data);
     }
 
     @Override
-    public void visit(BLangErrorVariableDef errorVariableDef) {
-        checkStatementExecutionValidity(errorVariableDef);
+    public void visit(BLangErrorVariableDef errorVariableDef, AnalyzerData data) {
+        checkStatementExecutionValidity(errorVariableDef, data);
         BLangExpression expr = errorVariableDef.errorVariable.expr;
-        analyzeReachabilityInExpressionIfApplicable(expr, env);
+        analyzeReachabilityInExpressionIfApplicable(expr, data);
     }
 
     @Override
-    public void visit(BLangFail failNode) {
-        checkStatementExecutionValidity(failNode);
-        this.errorThrown = true;
-        if (!this.failureHandled) {
-            this.statementReturnsPanicsOrFails = true;
+    public void visit(BLangFail failNode, AnalyzerData data) {
+        checkStatementExecutionValidity(failNode, data);
+        data.errorThrown = true;
+        if (!data.failureHandled) {
```

Comment:
Can't we propagate `failureHandled ` from CodeAnalyzer since these checks are being done during code analysis as well?  If so; since https://github.com/ballerina-platform/ballerina-lang/pull/34909/ is not merged yet; shall we create an issue to improve these kinds of scenarios?

Reply:
The reachability analysis is carried out after the code analysis in a new context. So we can't really propogate the statuses we keep in the code analyzer.